Futures

spall opened the pull request:
Here is an implementation of futures along with some changes to engines to make them thread safe.

There is a scheduler which maintains the threads that run the futures.  And each of these threads is running a simple round-robin scheduler that uses engines to control how long each future is run for.  There is also some very basic work stealing between threads.

I look forward to your feedback

mflatt commented:
Looks plausible.

To make it ready to merge – which I assume makes it easier for you to continue working – can you make it build in a non-threaded Chez Scheme? I imagine you'll need a `meta-cond` to pick between `make-parameter` and `make-thread-parameter`. (I see that `threaded?` is the predicate I wanted and didn't find for the `meta-cond` in "lock.ss".)

FWIW, I'd also be happier merging something that doesn't have a `sleep` call. It should be easy to gate the future workers using a condition (in the sense of `make-condition`) instead of a busy-sleeping loop.

spall commented:
I put a meta-cond in parameter.ss that selects between make-thread-parameter and make-parameter; and then just replaced all (if (threaded?) ...) with calls to   the new function defined in parameter.ss.

I removed the sleep call and replaced it with a condition-wait; definitely a better way of doing things.

spall commented:
Had to rebase after fixing a conflict on github then needing to push a minor repair.
